README.md
# DSA Performance Comparison

Performance comparison among different messaging protocols


CHANGELOG.md
# Changelog

## 0.8.4

- Added Kafka Example

## 0.8.3

- Add HiveMQ Example

## 0.8.2

- Convert MPS calculation to be over duration than between messages

## 0.8.1

- Fix to prevent empty hash

## 0.8.0

- Use start button to trigger loading and provide status.

## 0.7.1

- Fix issue with streaming starting before all examples loaded

## 0.7.0

- Report messages received per second (averaged over 10 messages).

## 0.6.0

* Trigger events so all examples have time to initialize first, then
they can all start sending events at the same time.

## 0.5.2

- Balancing received data with transmitted if not currently transmitting.

## 0.5.1

- Fix bug in PubNub sample not starting on new session.
- Use a data store for DSA instead of storage link.

## 0.5.0

- Use sessions now (via location hash). If no session is provided
a new session is created automagically.

## 0.4.8

- Added delta values to metrics

## 0.4.7

- Add metrics to page

## 0.4.6

- Offset crossbar and dsa bubbles to give some spacial disparity

## 0.4.5

- Update Crossbar to add event listener like other samples

## 0.4.4

- Update QOS default values for MQTT and DSA

## 0.4.3

- Migrate crossbar to server.

## 0.4.2

- Start migrating to the server
- Bugfixes for MQTT sample.

## 0.4.0

- Added MQTT sample

## 0.3.0

- Added PubNub sample.

## 0.2.0

- Added crossbar.io sample

## 0.1.0

- Initial version with dsa.

## 0.0.1

- Initial version


web/main.dart
import "dart:html";

import "package:dslink/browser.dart";
import "package:dslink/nodes.dart";

LinkProvider link;

final Map<String, String> TRANSITION_TIMES = {
  "Instant": "0s",
  "250ms": "0.25s",
  "500ms": "0.5s",
  "1s": "1s",
  "2s": "2s"
};

Map<String, dynamic> DEFAULT_NODES = {
  "User_Agent": {
    r"$name": "User Agent",
    r"$type": "string",
    "?value": window.navigator.userAgent
  },
  "Page_Color": { // Page Background Color
    r"$name": "Page Color",
    r"$type": "color",
    r"$writable": "write",
    "?value": "blue"
  },
  "Page_Gradient": {
    r"$name": "Page Gradient",
    r"$type": "gradient",
    r"$writable": "write",
    "?value": "none"
  },
  "Page_Color_Transition_Time": {
    r"$name": "Page Color Transition Time",
    r"$type": TRANSITION_ENUM,
    r"$writable": "write",
    "?value": "1s"
  },
  "Text": { // Text Message
    r"$name": "Text",
    r"$type": "string",
    r"$writable": "write",
    "?value": "Hello World"
  },
  "Text_Size_Transition_Time": {
    r"$name": "Text Size Transition Time",
    r"$type": TRANSITION_ENUM,
    r"$writable": "write",
    "?value": "1s"
  },
  "Text_Color_Transition_Time": {
    r"$name": "Text Color Transition Time",
    r"$type": TRANSITION_ENUM,
    r"$writable": "write",
    "?value": "1s"
  },
  "Text_Color": { // Text Color
    r"$name": "Text Color",
    r"$type": "color",
    r"$writable": "write",
    "?value": "white"
  },
  "Text_Font": {
    r"$name": "Text Font",
    r"$type": buildEnumType([
      "Arial",
      "Arial Black",
      "Comic Sans MS",
      "Courier New",
      "Georgia",
      "Impact",
      "Tahoma",
      "Lucida Console",
      "Times New Roman",
      "Trebuchet MS",
      "Verdana"
    ]),
    r"$writable": "write",
    "?value": "Arial"
  },
  "Text_Rotation": {
    r"$name": "Text Rotation",
    r"$type": "number",
    r"$writable": "write",
    "?value": 0.0
  },
  "Text_Size": {
    r"$name": "Text Size",
    r"$type": "number",
    r"$writable": "write",
    "?value": 72
  },
  "Click": {
    "ID": {
      r"$type": "number",
      "?value": 0
    },
    "X": {
      r"$type": "number",
      "?value": 0.0
    },
    "Y": {
      r"$type": "number",
      "?value": 0.0
    }
  },
  "Text_Hovering": { // If the user is currently hovering over the text.
    r"$name": "Hovering over Text",
    r"$type": "bool",
    "?value": false
  },
  "Mouse": { // Mouse-related stuff.
    "X": { // Mouse X position.
      r"$type": "number",
      "?value": 0.0
    },
    "Y": { // Mouse y position.
      r"$type": "number",
      "?value": 0.0
    },
    "Down": {
      r"$type": "bool",
      "?value": false
    }
  },
  "Play_Sound": { // An action to play a sound.
    r"$name": "Play Sound",
    r"$is": "playSound",
    r"$invokable": "write",
    r"$params": [
      {
        "name": "url",
        "type": "string"
      }
    ]
  },
  "Stop_Sound": {
    r"$name": "Stop Sound",
    r"$is": "stopSound",
    r"$invokable": "write"
  }
};

final String TRANSITION_ENUM = buildEnumType(TRANSITION_TIMES.keys);

BodyElement bodyElement = querySelector("#body");
ParagraphElement textElement = querySelector("#text");

AudioElement audio;

main() async {
  var brokerUrl = await BrowserUtils.fetchBrokerUrlFromPath("broker_url", "http://localhost:8080/conn");
  link = new LinkProvider(brokerUrl, "Browser-", defaultNodes: DEFAULT_NODES, profiles: {
    "playSound": (String path) =>
    new SimpleActionNode(path, (Map<String, dynamic> params) {
      if (audio != null) {
        audio.pause();
        audio = null;
      }

      audio = new AudioElement();
      audio.src = params["url"];
      audio.play();
    }),
    "stopSound": (String path) =>
    new SimpleActionNode(path, (Map<String, dynamic> params) {
      if (audio != null) {
        audio.pause();
        audio = null;
      }
    })
  });

  await $.init();

  for (var key in DEFAULT_NODES.keys) {
    if (!link["/"].children.containsKey(key)) {
      $.addNode(key, DEFAULT_NODES[key]);
    }
  }

  $.onValueChange("/Page_Color").listen((ValueUpdate update) async { // Wait for background color changes.
    String color = update.value;
    try {
      color = "#${int.parse(update.value).toRadixString(16)}";
    } catch (e) {}
    bodyElement
      ..style.removeProperty("background")
      ..style.backgroundColor = color;
    await $.save();
  });

  $.onValueChange("/Text_Color").listen((ValueUpdate update) async { // Wait for text color changes.
    String color = update.value;
    try {
      color = "#${int.parse(update.value).toRadixString(16)}";
    } catch (e) {}
    textElement
      ..style.color = color
      ..offsetHeight; // Trigger Re-flow
    await $.save();
  });

  $.onValueChange("/Page_Gradient").listen((ValueUpdate update) async {
    if (update.value == "none") {
      return;
    }

    if (update.value == null) {
      $.val("/Page_Gradient", "none");
      return;
    }

    String x = update.value;
    bodyElement
      ..style.removeProperty("background-color")
      ..style.background = "linear-gradient(${x})";
  });

  $.onValueChange("/Text").listen((ValueUpdate update) async { // Wait for message changes.
    textElement
      ..text = update.value
      ..offsetHeight; // Trigger Re-flow
    await $.save();
  });

  $.onValueChange("/Text_Rotation").listen((ValueUpdate update) async {
    textElement
      ..style.transform = "rotate(${update.value}deg";
    await $.save();
  });

  $.onValueChange("/Page_Color_Transition_Time").listen((ValueUpdate update) async {
    var n = update.value;

    if (TRANSITION_TIMES.containsKey(n)) {
      n = TRANSITION_TIMES[n];
    }

    bodyElement
      ..style.transition = "background-color ${n}";

    await $.save();
  });

  $.onValueChange("/Text_Font").listen((ValueUpdate update) async {
    textElement
      ..style.fontFamily = '"' + update.value + '"'
      ..offsetHeight;
    await $.save();
  });

  $.onValueChange("/Text_Size").listen((ValueUpdate update) async {
    textElement
      ..style.fontSize = "${update.value}px"
      ..offsetHeight;
    await $.save();
  });

  $.onValueChange("/Page_Color_Transition_Time").listen((ValueUpdate update) async {
    var n = update.value;

    if (TRANSITION_TIMES.containsKey(n)) {
      n = TRANSITION_TIMES[n];
    }

    bodyElement
      ..style.transition = "background-color ${n}";

    await $.save();
  });

  $.onValueChange("/Text_Size_Transition_Time").listen((ValueUpdate update) async {
    var n = update.value;

    if (TRANSITION_TIMES.containsKey(n)) {
      n = TRANSITION_TIMES[n];
    }

    textElement
      ..style.transition = "font-size ${n}";

    await $.save();
  });

  $.onValueChange("/Text_Color_Transition_Time").listen((ValueUpdate update) async {
    var n = update.value;

    if (TRANSITION_TIMES.containsKey(n)) {
      n = TRANSITION_TIMES[n];
    }

    textElement
      ..style.transition = "color ${n}";

    await $.save();
  });

  bodyElement.onClick.listen((MouseEvent event) { // Update Click Information
    $.val("/Click/ID", (link["/Click/ID"].lastValueUpdate.value as int) + 1);
    $.val("/Click/X", event.page.x);
    $.val("/Click/Y", event.page.y);
  });

  textElement.onMouseEnter.listen((event) => $.val("/Text_Hovering", true));
  textElement.onMouseLeave.listen((event) => $.val("/Text_Hovering", false));

  bodyElement.onMouseMove.listen((MouseEvent event) {
    $.val("/Mouse/X", event.page.x);
    $.val("/Mouse/Y", event.page.y);
  });

  bodyElement.onMouseDown.listen((_) {
    $.val("/Mouse/Down", true);
  });

  bodyElement.onMouseUp.listen((_) {
    $.val("/Mouse/Down", false);
  });

  // Re-sync Values to trigger subscribers.
  [
    "/Page_Color_Transition_Time",
    "/Text_Color_Transition_Time",
    "/Text_Size_Transition_Time",
    "/Page_Color",
    "/Page_Gradient",
    "/Text_Color",
    "/Text",
    "/Text_Font",
    "/Text_Size"
  ].forEach($.syncValue);

  $.connect();
}

LinkProvider get $ => link;

web/example.dart
library dsa.example;

import "dart:html";
import 'dart:async';
import 'dart:math';

import "package:dslink/browser.dart";
import "package:crypto/crypto.dart";

/// BodyElement to dispatch and listen to events.
final BodyElement BODY = querySelector('body');

/// Class manages unique session IDs and checks that all
/// samples are loaded before signaling the beginning of the tests
class ManageSession {
  static const TOTAL_TRIALS = 6;
  final MAX_SIZE = 10000;
  final LIST_SIZE = 10;
  String _hash;
  StorageExample se;
  ButtonElement _start;

  int _messagesReceived = 0;

  ManageSession() {
    _start = document.querySelector('#start');
    _start.onClick.listen((_) {
      _start..disabled = true
        ..text = 'Loading...';
      BODY.on['trialready'].listen(prepTrials);
      BODY.on['inittrials'].listen((event) {
        se = new StorageExample(_hash);
        se.init();
      });
      BODY.dispatchEvent(new CustomEvent('inittrials'));
    });
    _hash = window.location.hash;
    if(_hash.isEmpty || _hash.substring(1).isEmpty) {
      var list = new List<int>(LIST_SIZE);
      var rnd = new Random(new DateTime.now().millisecond);
      for(var i = 0; i < LIST_SIZE; i++) {
        list[i] = rnd.nextInt(MAX_SIZE);
      }
      _hash = CryptoUtils.bytesToHex(list);
      window.location.hash = _hash;
    } else {
      _hash = Uri.encodeQueryComponent(_hash.substring(1));
    }
  }

  void prepTrials(CustomEvent e) {
    _messagesReceived += 1;
    print('Ready from: ${e.detail['trial']}');
    if(_messagesReceived >= TOTAL_TRIALS) {
      BODY.dispatchEvent(new CustomEvent('appready'));
      _start.text = 'Running';
    }
  }
}

/// DSA sample using data nodes to store values and retreive them.
class StorageExample {
  static const PER_SECOND = 1000.0;
  static const TIME_TO_COUNT = const Duration(seconds: 5);
  DivElement _circle;
  SpanElement _txSpan;
  SpanElement _rxSpan;
  SpanElement _dtSpan;
  SpanElement _mpsSpan;
  num _tx;
  num _rx;
  num _lastMessageTime;
  LinkProvider link;
  Requester _req;
  String hash;
  String _node;
  Timer _msgTimer;

  StorageExample(this.hash) {
    _node = '/data/follow/$hash';
    _circle = querySelector('#dgcircle');
    _txSpan = querySelector('#dgmettx');
    _rxSpan = querySelector('#dgmetrx');
    _dtSpan = querySelector('#dgmetdt');
    _mpsSpan = querySelector('#dgmps');
    _tx = 0;
    _rx = 0;
    _lastMessageTime = 0.0;
  }

  Future init() async {
    var brokerUrl = await BrowserUtils.fetchBrokerUrlFromPath('broker_url',
        'http://performance.iot-dsa.org:8080/conn');
    link = new LinkProvider(brokerUrl,
        'Browser-', isResponder: false);

    await link.connect();
    _req = await link.onRequesterReady;
    _req.subscribe(_node, bothUpdated, 3);

    // Listen for start event before sending messages
    BODY.on['appready'].listen((_) {
      print('AppReady Received from: DSA');
      _msgTimer = new Timer.periodic(TIME_TO_COUNT, countMessages);
      BODY.onMouseMove.listen(mouseMoved);
    });
    BODY.dispatchEvent(
        new CustomEvent('trialready', detail: {'trial' : 'dsa'}));
  }

  void countMessages(Timer _) {
    var delta = _rx - _lastMessageTime;
    _lastMessageTime = _rx;
    _mpsSpan.text = '${delta / 5}';
  }

  void bothUpdated(ValueUpdate update) {
    _circle.style.left = '${(update.value as Map)['x']}px';
    _circle.style.top = '${(update.value as Map)['y'] - 30}px';
    _rxSpan.text = '${++_rx}';
    if(_rx > _tx) {
      _tx = _rx;
      _txSpan.text = '$_tx';
    }
    _dtSpan.text = '${_tx - _rx}';

//    if(_lastMessageTime == 0.0) {
//      _lastMessageTime = window.performance.now();
//    }

//    if(_rx % 10 == 0){
//      var now = window.performance.now();
//      var delta = now - _lastMessageTime;
//      _lastMessageTime = now;
//      _mpsSpan.text = (PER_SECOND / delta * 10).toStringAsFixed(2);
//    }
  }

  void mouseMoved(MouseEvent event) {
    var pos = {'x' : event.page.x, 'y': event.page.y };
    _req.set(_node, pos);
    _txSpan.text = '${++_tx}';
    _dtSpan.text = '${_tx - _rx}';
  }
}

main() async {
  var ms = new ManageSession();
}

web/hive-example.js
var HV_BODY = document.querySelector('body');

var HiveExample = (function() {

    var hvx;
    function HiveExample() {
        hvx = this;
        this.Elements = {};
        this.Elements.circle = document.getElementById('hvcircle');
        this.Elements.txSpan = document.getElementById('hvmettx');
        this.Elements.rxSpan = document.getElementById('hvmetrx');
        this.Elements.dtSpan = document.getElementById('hvmetdt');
        this.Elements.mps = document.getElementById('hvmps');
        this.rx = 0;
        this.tx = 0;
        this.lastMsgTime = 0;
        this.ssid = '';
        this.connection = null;
        this.PER_SECOND = 1000;

        var hash = window.location.hash;
        if(!hash) {
            window.addEventListener('hashchange', this.init);
        } else {
            this.init();
        }
    }

    HiveExample.prototype.mouseMoved = function(event) {
        var pos = {x: event.pageX, y: event.pageY};
        var message = new Paho.MQTT.Message(JSON.stringify(pos));
        message.destinationName = '/com.dglux.followme' + hvx.ssid;
        message.qos = 0;
        hvx.connection.send(message);
        hvx.Elements.txSpan.textContent = (hvx.tx += 1);
        hvx.Elements.dtSpan.textContent = (hvx.tx - hvx.rx);
    };

    HiveExample.prototype.messageCount = function() {
        var delta = hvx.rx - hvx.lastMsgTime;
        hvx.lastMsgTime = hvx.rx;
        hvx.Elements.mps.textContent = (delta / 5);
    };

    HiveExample.prototype.clientConnected = function() {
        hvx.connection.subscribe('/com.dglux.followme'+ hvx.ssid);

        HV_BODY.addEventListener('appready', function() {
            console.log('AppReady Received from: HiveMQ');
            window.setInterval(hvx.messageCount, 5000);
            HV_BODY.addEventListener('mousemove', hvx.mouseMoved);
        });
        HV_BODY.dispatchEvent(new CustomEvent('trialready', { detail: {trial: 'hive'}}));
    };

    HiveExample.prototype.messageReceived = function(msg) {
        var pos;
        try {
            pos = JSON.parse(msg.payloadString);
        } catch (e) {
            console.log('error:', e);
        }
        hvx.Elements.rxSpan.textContent = (hvx.rx += 1);
        if(hvx.rx > hvx.tx) {
            hvx.tx = hvx.rx;
            hvx.Elements.txSpan.textContent = hvx.tx;
        }
        hvx.Elements.dtSpan.textContent = (hvx.tx - hvx.rx);
        hvx.Elements.circle.style.left = pos.x + 'px';
        hvx.Elements.circle.style.top = pos.y + 'px';

        //if(hvx.lastMessageTime === 0) {
        //    hvx.lastMessageTime = window.performance.now();
        //}
        //
        //if(hvx.rx % 10 === 0) {
        //    var now = window.performance.now();
        //    var delta = now - hvx.lastMsgTime;
        //    hvx.lastMsgTime = now;
        //    hvx.Elements.mps.textContent = ((hvx.PER_SECOND / delta) * 10).toFixed(2);
        //}
    };

    HiveExample.prototype.init = function() {
        hvx.ssid = window.location.hash.substr(1);

        hvx.connection = new Paho.MQTT.Client('performance.iot-dsa.org', 8105, 'followme-1' + new Date().getTime());
        hvx.connection.onMessageArrived = hvx.messageReceived;
        hvx.connection.onConnectionLost = function onConnectionLost(responseObject) {
            if (responseObject.errorCode !== 0) {
                console.log("onConnectionLost:" + responseObject.errorMessage);
                console.log(responseObject);
            }
        };

        hvx.connection.connect({onSuccess: hvx.clientConnected});
    };

    return HiveExample;
})();

HV_BODY.addEventListener('inittrials', function () {
    var hvExample = new HiveExample();
});


web/mqtt-example.js
var MQ_BODY = document.querySelector('body');

var MQExample = (function() {

    var mqx;
    function MQExample() {
        mqx = this;
        this.Elements = {};
        this.Elements.circle = document.getElementById('mqttcircle');
        this.Elements.txSpan = document.getElementById('mqmettx');
        this.Elements.rxSpan = document.getElementById('mqmetrx');
        this.Elements.dtSpan = document.getElementById('mqmetdt');
        this.Elements.mps = document.getElementById('mqmps');
        this.rx = 0;
        this.tx = 0;
        this.lastMsgTime = 0;
        this.ssid = '';
        this.connection = null;
        this.PER_SECOND = 1000;

        var hash = window.location.hash;
        if(!hash) {
            window.addEventListener('hashchange', this.init);
        } else {
            this.init();
        }
    }

    MQExample.prototype.mouseMoved = function(event) {
        var pos = {x: event.pageX, y: event.pageY};
        var message = new Paho.MQTT.Message(JSON.stringify(pos));
        message.destinationName = '/com.dglux.followme' + mqx.ssid;
        message.qos = 0;
        mqx.connection.send(message);
        mqx.Elements.txSpan.textContent = (mqx.tx += 1);
        mqx.Elements.dtSpan.textContent = (mqx.tx - mqx.rx);
    };

    MQExample.prototype.messageCount = function() {
        var delta = mqx.rx - mqx.lastMsgTime;
        mqx.lastMsgTime = mqx.rx;
        mqx.Elements.mps.textContent = (delta / 5);
    };

    MQExample.prototype.clientConnected = function() {
        mqx.connection.subscribe('/com.dglux.followme'+ mqx.ssid);

        MQ_BODY.addEventListener('appready', function() {
            console.log('AppReady Received from: mosquito');
            window.setInterval(mqx.messageCount, 5000);
            MQ_BODY.addEventListener('mousemove', mqx.mouseMoved);
        });
        MQ_BODY.dispatchEvent(new CustomEvent('trialready', { detail: {trial: 'mosquito'}}));
    };

    MQExample.prototype.messageReceived = function(msg) {
        var pos;
        try {
            pos = JSON.parse(msg.payloadString);
        } catch (e) {
            console.log('error:', e);
        }
        mqx.Elements.rxSpan.textContent = (mqx.rx += 1);
        if(mqx.rx > mqx.tx) {
            mqx.tx = mqx.rx;
            mqx.Elements.txSpan.textContent = mqx.tx;
        }
        mqx.Elements.dtSpan.textContent = (mqx.tx - mqx.rx);
        mqx.Elements.circle.style.left = pos.x + 'px';
        mqx.Elements.circle.style.top = pos.y + 'px';

        //if(mqx.lastMessageTime === 0) {
        //    mqx.lastMessageTime = window.performance.now();
        //}
        //
        //if(mqx.rx % 10 === 0) {
        //    var now = window.performance.now();
        //    var delta = now - mqx.lastMsgTime;
        //    mqx.lastMsgTime = now;
        //    mqx.Elements.mps.textContent = ((mqx.PER_SECOND / delta) * 10).toFixed(2);
        //}
    };

    MQExample.prototype.init = function() {
        mqx.ssid = window.location.hash.substr(1);

        mqx.connection = new Paho.MQTT.Client('performance.iot-dsa.org', 8091, 'followme-1' + new Date().getTime());
        mqx.connection.onMessageArrived = mqx.messageReceived;
        mqx.connection.onConnectionLost = function onConnectionLost(responseObject) {
            if (responseObject.errorCode !== 0) {
                console.log("onConnectionLost:" + responseObject.errorMessage);
                console.log(responseObject);
            }
        };

        mqx.connection.connect({onSuccess: mqx.clientConnected});
    };

    return MQExample;
})();

MQ_BODY.addEventListener('inittrials', function () {
    var mqExample = new MQExample();
});


web/kafka-example.js
// PORT: 7080
var KF_BODY = document.querySelector('body');

var KafkaExample = (function() {

    var kfx;
    function KafkaExample() {
        kfx = this;
        this.Elements = {};
        this.Elements.circle = document.getElementById('kfcircle');
        this.Elements.txSpan = document.getElementById('kfmettx');
        this.Elements.rxSpan = document.getElementById('kfmetrx');
        this.Elements.dtSpan = document.getElementById('kfmetdt');
        this.Elements.mps = document.getElementById('kfmps');
        this.rx = 0;
        this.tx = 0;
        this.lastMsgTime = 0;
        this.ssid = '';
        this.connection = null;
        this.PER_SECOND = 1000;

        var hash = window.location.hash;
        if(!hash) {
            window.addEventListener('hashchange', this.init);
        } else {
            this.init();
        }
    }

    KafkaExample.prototype.messageCount = function(event) {
        var delta = kfx.rx - kfx.lastMsgTime;
        kfx.lastMsgTime = kfx.rx;
        kfx.Elements.mps.textContent = (delta / 5);
    };

    KafkaExample.prototype.mouseMoved = function(event) {
        var pos = JSON.stringify({x: event.pageX, y: event.pageY});
        var message = {topic: kfx.ssid, message: pos};
        kfx.connection.send(JSON.stringify(message));
        kfx.Elements.txSpan.textContent = (kfx.tx += 1);
        kfx.Elements.dtSpan.textContent = (kfx.tx - kfx.rx);
    };

    KafkaExample.prototype.clientConnected = function(event) {
        KF_BODY.addEventListener('appready', function() {
            console.log('AppReady Received from: mosquito');
            window.setInterval(kfx.messageCount, 5000);
            KF_BODY.addEventListener('mousemove', kfx.mouseMoved);
        });
        KF_BODY.dispatchEvent(new CustomEvent('trialready', { detail: {trial: 'Kafka'}}));
    };

    KafkaExample.prototype.messageReceived = function(event) {
        var msg, pos;
        try {
            msg = JSON.parse(event.data);
            pos = JSON.parse(msg.message);
        } catch (e) {
            console.log(e);
        }

        kfx.Elements.rxSpan.textContent = (kfx.rx += 1);
        if(kfx.rx > kfx.tx) {
            kfx.tx = kfx.rx;
            kfx.Elements.txSpan.textContent = kfx.tx;
        }
        kfx.Elements.dtSpan.textContent = (kfx.tx - kfx.rx);
        kfx.Elements.circle.style.left = pos.x + 'px';
        kfx.Elements.circle.style.top = (pos.y - 15) + 'px';

    };

    KafkaExample.prototype.init = function() {
        kfx.ssid = window.location.hash.substr(1);

        kfx.connection = new WebSocket("ws://performance.iot-dsa.org:7080/v2/broker/?topics=" + kfx.ssid);
        kfx.connection.onmessage = kfx.messageReceived;
        kfx.connection.onopen = kfx.clientConnected;
    };

    return KafkaExample;
})();

KF_BODY.addEventListener('inittrials', function () {
    var kfExample = new KafkaExample();
});

web/pubnub-example.js
var PN_BODY = document.querySelector('body');

var PNExample = (function() {

    var pnx;
    function PNExample() {
        pnx = this;
        this.rx = 0;
        this.tx = 0;
        this.lastMsgTime = 0;
        this.Elements = {};
        this.Elements.circle = document.getElementById('pncircle');
        this.Elements.txSpan = document.getElementById('pnmettx');
        this.Elements.rxSpan = document.getElementById('pnmetrx');
        this.Elements.dtSpan = document.getElementById('pnmetdt');
        this.Elements.mps = document.getElementById('pnmps');
        this.connection = null;
        this.ssid = '';
        this.PER_SECOND = 1000;

        var hash = window.location.hash;
        if(!hash) {
            window.addEventListener('hashchange', this.init);
        } else {
            this.init();
        }
    }

    PNExample.prototype.messageReceived = function(data) {
        pnx.Elements.rxSpan.textContent = (pnx.rx += 1);
        if(pnx.rx > pnx.tx) {
            pnx.tx = pnx.rx;
            pnx.Elements.txSpan.textContent = pnx.tx;
        }
        pnx.Elements.dtSpan.textContent = (pnx.tx - pnx.rx);
        pnx.Elements.circle.style.left = data.x + 'px';
        pnx.Elements.circle.style.top = data.y + 'px';

        //if(pnx.lastMsgTime === 0) {
        //    pnx.lastMsgTime = window.performance.now();
        //}
        //
        //if(pnx.rx % 10 === 0) {
        //    var now = window.performance.now();
        //    var delta = now - pnx.lastMsgTime;
        //    pnx.lastMsgTime = now;
        //    pnx.Elements.mps.textContent = (pnx.PER_SECOND / delta * 10).toFixed(2);
        //}
    };

    PNExample.prototype.mouseMoved = function(event) {
        var pos = {x: event.pageX, y: event.pageY};
        pnx.connection.publish({
            channel: 'com.dglux.followme' + pnx.ssid,
            message: pos
        });
        pnx.Elements.txSpan.textContent = (pnx.tx += 1);
        pnx.Elements.dtSpan.textContent = (pnx.tx - pnx.rx);
    };

    PNExample.prototype.messageCount = function() {
        var delta = pnx.rx - pnx.lastMsgTime;
        pnx.lastMsgTime = pnx.rx;
        pnx.Elements.mps.textContent = (delta / 5);
    }

    PNExample.prototype.init = function() {
        pnx.ssid = window.location.hash.substr(1);
        pnx.connection = PUBNUB.init({
            publish_key: 'pub-c-85f4ff98-5adc-4c44-a915-90efb914f56e',
            subscribe_key: 'sub-c-2a0a5cb2-6228-11e5-bba9-02ee2ddab7fe'
        });

        pnx.connection.subscribe({
            channel: 'com.dglux.followme' + pnx.ssid,
            message: pnx.messageReceived
        });
        PN_BODY.addEventListener('appready', function() {
            console.log('AppReady Received from: PubNub');
            window.setInterval(pnx.messageCount, 5000);
            PN_BODY.addEventListener('mousemove', pnx.mouseMoved);
        });
        PN_BODY.dispatchEvent(new CustomEvent('trialready', { detail: {trial: 'pubnub'}}));
    };

    return PNExample;
})();

CB_BODY.addEventListener('inittrials', function () {
    var pnExample = new PNExample();
});

web/crossbar-example.js
var CB_BODY = document.querySelector('body');

var CrossbarExample = (function() {
    var cbx;
    function CrossbarExample() {
        cbx = this;
        // the URL of the WAMP Router (Crossbar.io)
        this.wsuri = "ws://performance.iot-dsa.org:8095/ws";
        this.tx = 0;
        this.rx = 0;
        this.lastMsgTime = 0;
        this.Elements = {};
        this.Elements.circle = document.getElementById('cbcircle');
        this.Elements.txSpan = document.getElementById('cbmettx');
        this.Elements.rxSpan = document.getElementById('cbmetrx');
        this.Elements.dtSpan = document.getElementById('cbmetdt');
        this.Elements.mps = document.getElementById('cbmps');
        this.rxConnection = null;
        this.txConnection = null;
        this.ssid = '';
        this.PER_SECOND = 1000;

        var hash = window.location.hash;
        if(!hash) {
            window.addEventListener('hashchange', this.init);
        } else {
            this.init();
        }
    }

    CrossbarExample.prototype.init = function() {
        var hash = window.location.hash.substr(1);
        cbx.rxConnection = new autobahn.Connection({ url: cbx.wsuri, realm: "realm1" });
        cbx.txConnection = new autobahn.Connection({ url: cbx.wsuri, realm: "realm1" });

        cbx.ssid = 'com.dglogik.circles-' + hash;

        cbx.rxConnection.onopen = function (session, details) {
            // SUBSCRIBE to a topic and receive events
            function on_updated(args, kwargs, details) {
                var pos = args[0];
                cbx.Elements.circle.style.left = pos.x + 'px';
                cbx.Elements.circle.style.top = (pos.y + 10) + 'px';
                cbx.Elements.rxSpan.textContent = (cbx.rx += 1);
                if(cbx.rx > cbx.tx) {
                    cbx.tx = cbx.rx;
                    cbx.Elements.txSpan.textContent = cbx.tx;
                }
                cbx.Elements.dtSpan.textContent = (cbx.tx - cbx.rx);

                //if(cbx.lastMsgTime === 0) {
                //    cbx.lastMsgTime = window.performance.now();
                //}
                //
                //if(cbx.rx % 10 === 0) {
                //    var now = window.performance.now();
                //    var delta = now - cbx.lastMsgTime;
                //    cbx.lastMsgTime = now;
                //    cbx.Elements.mps.textContent = ((cbx.PER_SECOND / delta) * 10).toFixed(2);
                //}
            }

            session.subscribe(cbx.ssid, on_updated).then(
                function (sub) {
                },
                function (err) {
                    console.log('failed to subscribe to topic: ', topic, err);
                }
            );
        };

        cbx.txConnection.onopen = function (session, details) {
            function mouseMoved(event) {
                var pos = {x: event.pageX, y: event.pageY};
                session.publish(cbx.ssid, [pos], {}, {acknowledge: true}).then(
                    function (pub) {
                    },
                    function (err) {
                        console.log('pub error:', err);
                    }
                );
                cbx.Elements.txSpan.textContent = (cbx.tx += 1);
                cbx.Elements.dtSpan.textContent = (cbx.tx - cbx.rx);
            }
            function messageCount() {
                var delta = cbx.rx - cbx.lastMsgTime;
                cbx.lastMsgTime = cbx.rx;
                cbx.Elements.mps.textContent = (delta / 5);
            }
            CB_BODY.addEventListener('appready', function() {
                console.log('AppReady Received from: Crossbar');
                CB_BODY.addEventListener('mousemove', mouseMoved);
                window.setInterval(messageCount, 5000);
            });

            CB_BODY.dispatchEvent(new CustomEvent('trialready', { detail: {trial: 'crossbar'}}));
        };

        cbx.rxConnection.open();
        cbx.txConnection.open();
    };

    return CrossbarExample;
})();

CB_BODY.addEventListener('inittrials', function () {
    var cbExample = new CrossbarExample();
});